Sales Product Analysis

In [1]:
import pandas as pd
import os
import numpy as np
import seaborn as sns
import warnings
from matplotlib import pylab as plt
from statsmodels.graphics.gofplots import qqplot
from IPython.core.interactiveshell import InteractiveShell

In [2]:
warnings.simplefilter(action="ignore", category=FutureWarning)
warnings.simplefilter(action="ignore", category=Warning)
InteractiveShell.ast_node_interactivity = 'all'

%reload_ext autoreload
%autoreload 2
%matplotlib inline

def set_seed(seed=42):
    np.random.seed(seed)
    os.environ["PYTHONHASHSEED"] = str(seed)
set_seed()

sns.set_style('whitegrid')
sns.set_context('paper', font_scale=1.5)

pd.set_option("display.width", 100)
pd.set_option("display.max_columns", 50)
pd.set_option("display.max_rows", 30)

plt.style.use('fivethirtyeight')
plt.rcParams["axes.grid"] = True
plt.rcParams["grid.alpha"] = 0.5
plt.rcParams["grid.linestyle"] = '--'
plt.rcParams["font.family"] = "monospace"
plt.rcParams["axes.facecolor"] = "white"
plt.rcParams['axes.edgecolor'] = 'black'
plt.rcParams['figure.frameon'] = True
plt.rcParams['axes.spines.left'] = True
plt.rcParams['axes.spines.bottom'] = True
plt.rcParams['axes.spines.top'] = False
plt.rcParams['axes.spines.right'] = False
plt.rcParams['axes.linewidth'] = 1.5

print("setup-complete!")

setup-complete!


Merge the 12 months of sales data into a single CSV file

In [6]:
# let's make a list compreension for all the data in the folder
files = [file for file in os.listdir('E:/AMotefaker/ABC/Data Analyst/Sales Data Analysis/Sales Product Analysis/DataSet/')] 
# let's make a pandas DataFrame
all_months_data = pd.DataFrame()
# makes a loop for concat the data
for file in files:
    data = pd.read_csv("E:/AMotefaker/ABC/Data Analyst/Sales Data Analysis/Sales Product Analysis/DataSet/" + file)
    all_months_data = pd.concat([all_months_data, data])
# export all data to csv    
all_months_data.to_csv("E:/AMotefaker/ABC/Data Analyst/Sales Data Analysis/Sales Product Analysis/all_data.csv", index=False)

Read in updated DataFrame

In [7]:
sales_data = pd.read_csv('all_data.csv') # read data
sales_data # show data

Order ID                     Product Quantity Ordered Price Each      Order Date  \
0        176558        USB-C Charging Cable                2      11.95  04/19/19 08:46   
1           NaN                         NaN              NaN        NaN             NaN   
2        176559  Bose SoundSport Headphones                1      99.99  04/07/19 22:30   
3        176560                Google Phone                1        600  04/12/19 14:38   
4        176560            Wired Headphones                1      11.99  04/12/19 14:38   
...         ...                         ...              ...        ...             ...   
186845   259353      AAA Batteries (4-pack)                3       2.99  09/17/19 20:56   
186846   259354                      iPhone                1        700  09/01/19 16:00   
186847   259355                      iPhone                1        700  09/23/19 07:39   
186848   259356      34in Ultrawide Monitor                1     379.99  09/19/19 17:30   
186849   259357        USB-C Charging Cable                1      11.95  09/30/19 00:18   

                               Purchase Address  
0                  917 1st St, Dallas, TX 75001  
1                                           NaN  
2             682 Chestnut St, Boston, MA 02215  
3          669 Spruce St, Los Angeles, CA 90001  
4          669 Spruce St, Los Angeles, CA 90001  
...                                         ...  
186845   840 Highland St, Los Angeles, CA 90001  
186846  216 Dogwood St, San Francisco, CA 94016  
186847     220 12th St, San Francisco, CA 94016  
186848   511 Forest St, San Francisco, CA 94016  
186849   250 Meadow St, San Francisco, CA 94016  

[186850 rows x 6 columns]

In [8]:
"Head"
sales_data.head() # Checking the first 5 rows of data
"Tail"
sales_data.tail() # Checking the last 5 rows of data

'Head'

Order ID                     Product Quantity Ordered Price Each      Order Date  \
0   176558        USB-C Charging Cable                2      11.95  04/19/19 08:46   
1      NaN                         NaN              NaN        NaN             NaN   
2   176559  Bose SoundSport Headphones                1      99.99  04/07/19 22:30   
3   176560                Google Phone                1        600  04/12/19 14:38   
4   176560            Wired Headphones                1      11.99  04/12/19 14:38   

                       Purchase Address  
0          917 1st St, Dallas, TX 75001  
1                                   NaN  
2     682 Chestnut St, Boston, MA 02215  
3  669 Spruce St, Los Angeles, CA 90001  
4  669 Spruce St, Los Angeles, CA 90001

'Tail'

Order ID                 Product Quantity Ordered Price Each      Order Date  \
186845   259353  AAA Batteries (4-pack)                3       2.99  09/17/19 20:56   
186846   259354                  iPhone                1        700  09/01/19 16:00   
186847   259355                  iPhone                1        700  09/23/19 07:39   
186848   259356  34in Ultrawide Monitor                1     379.99  09/19/19 17:30   
186849   259357    USB-C Charging Cable                1      11.95  09/30/19 00:18   

                               Purchase Address  
186845   840 Highland St, Los Angeles, CA 90001  
186846  216 Dogwood St, San Francisco, CA 94016  
186847     220 12th St, San Francisco, CA 94016  
186848   511 Forest St, San Francisco, CA 94016  
186849   250 Meadow St, San Francisco, CA 94016

Data Preprocessing

In [9]:
# getting the information 
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186850 entries, 0 to 186849
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          186305 non-null  object
 1   Product           186305 non-null  object
 2   Quantity Ordered  186305 non-null  object
 3   Price Each        186305 non-null  object
 4   Order Date        186305 non-null  object
 5   Purchase Address  186305 non-null  object
dtypes: object(6)
memory usage: 8.6+ MB


Uniqueness Categorical Variables

In [10]:
categorical = sales_data.select_dtypes(['category', 'object']).columns # getting the Uniqueness catrgorical variable
for col in categorical:
    print('{} : {} unique value(s)'.format(col, sales_data[col].nunique()))

Order ID : 178438 unique value(s)
Product : 20 unique value(s)
Quantity Ordered : 10 unique value(s)
Price Each : 24 unique value(s)
Order Date : 142396 unique value(s)
Purchase Address : 140788 unique value(s)


How many missing data points do we have?

In [11]:
# get the number of missing data points per column
missing_values_count = sales_data.isnull().sum()
# look at the # of missing points in the first ten columns
missing_values_count[0:10]

Order ID            545
Product             545
Quantity Ordered    545
Price Each          545
Order Date          545
Purchase Address    545
dtype: int64

what percentage of the values in our dataset were missing to give us a better sense of the scale of this problem

In [12]:
# how many total missing values do we have?
total_cells = np.product(sales_data.shape)
total_missing = missing_values_count.sum()
# percent of data that is missing
percent_missing = (total_missing / total_cells) * 100
print(f"{percent_missing:.2f}%")

0.29%


Clean up the Data

In [13]:
# let's drop the rows of NaN data!
sales_data = sales_data.dropna(how='all')
# okay, let's check it again!
"NaN Value:"
sales_data[sales_data.isna().any(axis=1)]
# future warning! ValueError: invalid literal for int() with base 10: 'Or'
"Clean Future Warnings:"
sales_data = sales_data[sales_data['Order Date'].str[0:2] != 'Or']
sales_data

'NaN Value:'

Empty DataFrame
Columns: [Order ID, Product, Quantity Ordered, Price Each, Order Date, Purchase Address]
Index: []

'Clean Future Warnings:'

Order ID                     Product Quantity Ordered Price Each      Order Date  \
0        176558        USB-C Charging Cable                2      11.95  04/19/19 08:46   
2        176559  Bose SoundSport Headphones                1      99.99  04/07/19 22:30   
3        176560                Google Phone                1        600  04/12/19 14:38   
4        176560            Wired Headphones                1      11.99  04/12/19 14:38   
5        176561            Wired Headphones                1      11.99  04/30/19 09:27   
...         ...                         ...              ...        ...             ...   
186845   259353      AAA Batteries (4-pack)                3       2.99  09/17/19 20:56   
186846   259354                      iPhone                1        700  09/01/19 16:00   
186847   259355                      iPhone                1        700  09/23/19 07:39   
186848   259356      34in Ultrawide Monitor                1     379.99  09/19/19 17:30   
186849   259357        USB-C Charging Cable                1      11.95  09/30/19 00:18   

                               Purchase Address  
0                  917 1st St, Dallas, TX 75001  
2             682 Chestnut St, Boston, MA 02215  
3          669 Spruce St, Los Angeles, CA 90001  
4          669 Spruce St, Los Angeles, CA 90001  
5             333 8th St, Los Angeles, CA 90001  
...                                         ...  
186845   840 Highland St, Los Angeles, CA 90001  
186846  216 Dogwood St, San Francisco, CA 94016  
186847     220 12th St, San Francisco, CA 94016  
186848   511 Forest St, San Francisco, CA 94016  
186849   250 Meadow St, San Francisco, CA 94016  

[185950 rows x 6 columns]

Convert Quantity Ordered column and Price Each column

In [14]:
# convert the data
sales_data['Quantity Ordered'], sales_data['Price Each'] = sales_data['Quantity Ordered'].astype('int64'), sales_data['Price Each'].astype('float')
# and check it 
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 185950 entries, 0 to 186849
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Order ID          185950 non-null  object 
 1   Product           185950 non-null  object 
 2   Quantity Ordered  185950 non-null  int64  
 3   Price Each        185950 non-null  float64
 4   Order Date        185950 non-null  object 
 5   Purchase Address  185950 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 9.9+ MB


Convert Order Date column

In [15]:
# convert it using to_datetime() funct
sales_data['Order Date'] = pd.to_datetime(sales_data['Order Date'])
# let's see it
sales_data

Order ID                     Product  Quantity Ordered  Price Each          Order Date  \
0        176558        USB-C Charging Cable                 2       11.95 2019-04-19 08:46:00   
2        176559  Bose SoundSport Headphones                 1       99.99 2019-04-07 22:30:00   
3        176560                Google Phone                 1      600.00 2019-04-12 14:38:00   
4        176560            Wired Headphones                 1       11.99 2019-04-12 14:38:00   
5        176561            Wired Headphones                 1       11.99 2019-04-30 09:27:00   
...         ...                         ...               ...         ...                 ...   
186845   259353      AAA Batteries (4-pack)                 3        2.99 2019-09-17 20:56:00   
186846   259354                      iPhone                 1      700.00 2019-09-01 16:00:00   
186847   259355                      iPhone                 1      700.00 2019-09-23 07:39:00   
186848   259356      34in Ultrawide Monitor                 1      379.99 2019-09-19 17:30:00   
186849   259357        USB-C Charging Cable                 1       11.95 2019-09-30 00:18:00   

                               Purchase Address  
0                  917 1st St, Dallas, TX 75001  
2             682 Chestnut St, Boston, MA 02215  
3          669 Spruce St, Los Angeles, CA 90001  
4          669 Spruce St, Los Angeles, CA 90001  
5             333 8th St, Los Angeles, CA 90001  
...                                         ...  
186845   840 Highland St, Los Angeles, CA 90001  
186846  216 Dogwood St, San Francisco, CA 94016  
186847     220 12th St, San Francisco, CA 94016  
186848   511 Forest St, San Francisco, CA 94016  
186849   250 Meadow St, San Francisco, CA 94016  

[185950 rows x 6 columns]

Recap Data:

We have total  186850  records and  6  columns cateogircal type

The total of missing value that we have is  0.29167 %

Order ID :  178438  unique value(s)

Product :  20  unique value(s)

Quantity Ordered :  10  unique value(s)

Price Each :  24  unique value(s)

Order Date :  142396  unique value(s)

Purchase Address :  140788  unique value(s)

### Data Preparation

Add Month, Hour, Minute, Sales, Cities Column

In [16]:
def augment_data(data):
    
    """
    Adding new features to
    our data, adding Month Data,
    Hour Data, Minute Data, Sales Data,
    and Cities Column
    
    Returning:
        data with new features
    """
    
    # funtction to get the city in the data
    def get_city(address):
        return address.split(',')[1]
    
    # funtction to get the state in the data
    def get_state(address):
        return address.split(',')[2].split(' ')[1]

    # let's get the year data in order date column
    data['Year'] = data['Order Date'].dt.year
    
    # let's get the month data in order date column
    data['Month'] = data['Order Date'].dt.month
    
    # let's get the houe data in order date column
    data['Hour'] = data['Order Date'].dt.hour 
    
    # let's get the minute data in order date column
    data['Minute'] = data['Order Date'].dt.minute 
    
    # let's make the sales column by multiplying the quantity ordered colum with price each column
    data['Sales'] = data['Quantity Ordered'] * data['Price Each'] 
    
    # let's get the cities data in order date column
    data['Cities'] = data['Purchase Address'].apply(lambda x: f"{get_city(x)} ({get_state(x)})") 
    
    return data # returning data

# and see it
sales_data = augment_data(sales_data)
sales_data.head()

Order ID                     Product  Quantity Ordered  Price Each          Order Date  \
0   176558        USB-C Charging Cable                 2       11.95 2019-04-19 08:46:00   
2   176559  Bose SoundSport Headphones                 1       99.99 2019-04-07 22:30:00   
3   176560                Google Phone                 1      600.00 2019-04-12 14:38:00   
4   176560            Wired Headphones                 1       11.99 2019-04-12 14:38:00   
5   176561            Wired Headphones                 1       11.99 2019-04-30 09:27:00   

                       Purchase Address  Year  Month  Hour  Minute   Sales             Cities  
0          917 1st St, Dallas, TX 75001  2019      4     8      46   23.90        Dallas (TX)  
2     682 Chestnut St, Boston, MA 02215  2019      4    22      30   99.99        Boston (MA)  
3  669 Spruce St, Los Angeles, CA 90001  2019      4    14      38  600.00   Los Angeles (CA)  
4  669 Spruce St, Los Angeles, CA 90001  2019      4    14      38   11.99   Los Angeles (CA)  
5     333 8th St, Los Angeles, CA 90001  2019      4     9      27   11.99   Los Angeles (CA)